Add templates for presentation viewer

PrajwolAmatya opened the pull request:
<!--
Thanks for submitting a change to web app presentation viewer!

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the main branch which holds the next version of web app presentation viewer.

Please set the following labels:

- Assignment: assign to self
- Reviewers: pick at least one
- Labels: pick at least one
- Project: Web App Presentation Viewer (JankariTech)
-->

## Description
This PR adds basic templates for the presentation viewer.

Following templates are added:
1. Cover Slide Template

<img width="1888" height="914" alt="image" src="https://github.com/user-attachments/assets/1669fdb1-c415-47b0-ae37-3db18af19e0c" />

2. Title Content Slide Template

<img width="1888" height="914" alt="image" src="https://github.com/user-attachments/assets/1a96fb0e-c33e-4880-bc27-cb98aeab1452" />

3. Title Content Image Slide Template

<img width="1888" height="914" alt="image" src="https://github.com/user-attachments/assets/2302eb87-7de4-4347-9c8e-17883fe225ba" />

4. About Us Slide Template

<img width="1888" height="914" alt="image" src="https://github.com/user-attachments/assets/e8ac44ca-deb4-4915-b8fe-55d59943e136" />


### Slide content fitting

The font size of the slide will be automatically reduced when the content tends to overflow out of the slide.

<img width="1888" height="914" alt="image" src="https://github.com/user-attachments/assets/26e3fc68-251d-4a0b-b9cf-ad8b0bb29793" />

saw-jan commented on README.md:
```diff
@@ -50,6 +50,208 @@ This app has the following default slide separators:
+
+Besides default metadata, you can also provide inline metadata for each slide. This will be useful when you want to
+override the default metadata for a specific slide and this will be applied to that slide only. The inline metadata can
+be defined by adding metadata directly after the slide title as below:
```
```suggestion
be defined by adding metadata directly on the slide title as below:
```

saw-jan commented on src/App.vue:
```diff
@@ -207,6 +233,172 @@ function dirname(path: string) {
+  }
+}
+function updateImageStructure() {
+  const pTags = document.querySelectorAll('p > img')
```
element selection without scope grabs the elements globally

saw-jan commented on src/App.vue:
```diff
@@ -207,6 +233,174 @@ function dirname(path: string) {
+  }
+}
+function fitContent() {
+  const images = document.querySelectorAll('img')
```
here as well. let's scope the selector when using `document.`

saw-jan commented on README.md:
```diff
@@ -50,6 +50,215 @@ This app has the following default slide separators:
+| `cover`               | This slide template can be used for cover slide                 | `::slide:cover`               |
+| `title-content`       | The slide with title and content (content can be image as well) | `::slide:title-content`       |
+| `title-content-image` | The slide with title, and content in left and image in right    | `::slide:title-content-image` |
+| `about-us`            | The slide with company info (e.g., Values, mission, Vision)     | `::slide:about-us`            |
```
```suggestion
| `about-us`            | The slide with about-us info (e.g., Values, mission, Vision)     | `::slide:about-us`            |
```

saw-jan commented on README.md:
```diff
@@ -50,6 +50,215 @@ This app has the following default slide separators:
+logo: <path-to-logo>
+color: <font-color-for-slide-title>
+footer: <footer-content>
+aboutUs: <content-for-about-us-slide>
```
where do I see this content. I do not see in the about-us slide
```md
---
slide: cover
presenter: Tester
logo: ./car.png
color: red
footer: my foot
aboutUs: this is about us
---

## Slide 1 ::slide:cover

content one

---
## Slide 2 ::slide:title-content

content two

---
## About Us ::slide:about-us

content two
```
<img width="1396" height="878" alt="Screenshot from 2025-10-31 10-45-09" src="https://github.com/user-attachments/assets/f575f25c-fed3-45d9-99e6-6d22cbfe2248" />

PrajwolAmatya replied:
Missed to update the documenatation here. it should be as below
```markdown
---
slide: <default-slide-type>
presenter: <name-of-the-presenter>
logo: <path-to-logo>
color: <font-color-for-slide-title>
footer: <footer-content>
aboutUs:
    - title: <about-us-title1>
      text: <text-for-title1>
    - title: <about-us-title2>
      text: <text-for-title2>
    - title: <about-us-title3>
      text: <text-for-title3>
---
```

the reason for doing this is we could not split the about-us contents into different columns in the templates using mustache.

In `revealjs-awesoMD` plugin, we send the content to the slide as
```js
tempDiv.innerHTML = Mustache.render(xhr.responseText, {
    title: title,
    content: slideContent,
    metadata: options.metadata,
})
```
Here title is the title of the slide and we send the content through the content variable.

For the following markdown:
```markdown
# Title of slide
hello world
- some
- content
```
here,
title = Title of slide
content = 
```
hello world
- some
- content
```

In the templates we have
```html
<div class="content">
<div class="about-us-text">

{{{ content }}}

</div>
<div class="info-section">

{{#metadata.aboutUs}}
<div class="info-box">
<h3>{{ title }}</h3>
<p>{{ text }}</p>
</div>
<div class="divider"></div>
{{/metadata.aboutUs}}

</div>

</div>
```
Here if we provide the column wise contents in the content variable, we cannot split the content column wise as the content to show in each column could be anything, which makes it difficult to differentiate between the content.
That's why we provide these data from the default metadata in the frontmatter.

PrajwolAmatya commented on README.md:
````diff
@@ -50,6 +50,215 @@ This app has the following default slide separators:
+
+```markdown
+---
+slide: <default-slide-type>
````
The slide variable is already defined in the plugin.

saw-jan reviewed (approved):
👍 ✨

individual-it commented on README.md:
```diff
@@ -50,6 +50,223 @@ This app has the following default slide separators:
+### Slide Templates
+
+The slide template should be provided in the default metadata. Provide slide template that will be mostly used in the
+default metadata, so you don't need to add the slide metadata in every slide.
```
This is confusing, I want to  have different templates for different slides. But the template that is mostly used should be set in the default. Is that what you want to say?

individual-it commented:
pnpm update and then make command did not help

individual-it commented on README.md:
````diff
@@ -50,6 +50,222 @@ This app has the following default slide separators:
+
+```markdown
+---
+slide: <default-slide-type>
````
not giving a default slide, breaks the presentation

PrajwolAmatya replied:
> not giving a default slide, breaks the presentation

That was because i only checked if slide is provided in the default metadata to load styling for the templates. I have fixed it.

> special characters in the slide-name e.g. `<title-content>` or `<title-content` break the presentation

It is because the slide type doesn't exist. So, it should not break the whole presentation. Just the slide where invalid slide type is defined, It currently shows blank slide.
We are supposed to provide valid slide name. It should work same as what we did in openproject. The slide name is actually checked in the plugin.

individual-it replied:
If I provide `slide: title-content-does-not-exist` in the default metadata I get

<img width="1783" height="28" alt="grafik" src="https://github.com/user-attachments/assets/c385f13d-9210-4de5-a4e6-0327ed7e7a44" />

:+1:

If I provide: `slide: <title-content>` I get
<img width="1089" height="448" alt="grafik" src="https://github.com/user-attachments/assets/db22b0d5-2ba8-4181-9c24-9108ac560f5a" />

Not sure what happens but would be good to see the same error message

individual-it reviewed (changes requested):
would be good to have a "blank" template that works just as the default reveal-js template

individual-it commented on README.md:
```diff
@@ -50,6 +50,222 @@ This app has the following default slide separators:
+footer: <footer-content>
+aboutUs:
+    - title: <about-us-title1>
+      text: <text-for-title1>
```
how difficult would it be to allow links here?
Currently
```
  - title: WWW
     text: [www.jankaritech.com](www.jankaritech.com)
```

makes the whole presentation to fall apart

```
Object { name: "YAMLException", reason: "bad indentation of a mapping entry", mark: {…}, message: "bad indentation of a mapping entry at line 7, column 10:\n         text: [www.jankaritech.com](www.jank ... \n             ^", stack: "", … }
​
mark: Object { buffer: "slide: title-content\npresenter:  JankariTech Pvt. Ltd. \nlogo: Logo-colour.png\nfooter: Markdown Presentation Viewer\naboutUs:\n  - title: WWW\n     text: [www.jankaritech.com](www.jankaritech.com)\n  - title: linkedin\n    text: www.linkedin.com/company/jankaritech\n  - title: facebook\n    text: www.facebook.com/jankaritech\n\u0000", position: 148, line: 6, … }
​
message: "bad indentation of a mapping entry at line 7, column 10:\n         text: [www.jankaritech.com](www.jank ... \n             ^"
​
name: "YAMLException"
​
reason: "bad indentation of a mapping entry"
```

If links are not supported that is fine, but the presentation should not break

PrajwolAmatya commented:
I have addressed the issues and fixed them. Most of the fixes needed to done in the plugin (PR: https://github.com/opf/revealjs-awesoMD/pull/24). 


> Let's add tests for the templating to make sure all these corner cases don't break

Will work on it. :+1: 

@individual-it

individual-it reviewed (changes requested):
anything shown to the user e.g. error messages should be sanitized.

currently I can do:
```
# This template does not exist and has special char  ::slide:<a href="jankari.tech">jankaritech</a>
```

and the browser will render `<a href="jankari.tech">jankaritech</a>` as HTML

individual-it reviewed (changes requested):
sometimes the logo works sometimes not

I am using
```
---
slide: title-content
presenter:  JankariTech Pvt. Ltd. 
logo: images/Logo-colour.png
```
and the file `images/Logo-colour.png` does exist.

the URL of the logo in the presentation is sometimes: https://localhost:9200/images/Logo-colour.png but after a reload correctly `blob:https://localhost:9200/9db3f52b-2d67-4be4-b116-9a94f6798c7`

PrajwolAmatya commented:
> anything shown to the user e.g. error messages should be sanitized.
> 
> currently I can do:
> 
> ```
> # This template does not exist and has special char  ::slide:<a href="jankari.tech">jankaritech</a>
> ```
> 
> and the browser will render `<a href="jankari.tech">jankaritech</a>` as HTML

Fixed it in PR https://github.com/opf/revealjs-awesoMD/pull/25.
<img width="1206" height="883" alt="Screenshot from 2025-11-20 10-59-52" src="https://github.com/user-attachments/assets/ea459e7c-6e2a-4289-a978-926a3878e257" />
